### PR TITLE
[lake/iceberg] Add iceberg it case

### LINKED ISSUE #1564: Add integrating test for iceberg
### Search before asking

- [x] I searched in the [issues](https://github.com/alibaba/fluss/issues) and found nothing similar.

### Description

Just as `PaimonTieringITCase`, we should also add ITcase for iceberg.

### Willingness to contribute

- [x] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by luoyuxia**
### Purpose

Linked issue: close #1564

### CODE REVIEW

File: fluss-lake/fluss-lake-iceberg/src/test/java/com/alibaba/fluss/lake/iceberg/tiering/IcebergTieringITCase.java [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+    @BeforeAll
+    protected static void beforeAll() {
+        FlinkIcebergTieringTestBase.beforeAll();
+        execEnv = StreamExecutionEnvironment.getExecutionEnvironment();
+        execEnv.setParallelism(2);
+        execEnv.enableCheckpointing(1000);
+    }
+
+    @Test
+    void testTiering() throws Exception {
```
Comment (MehulBatra) [inline fluss-lake/fluss-lake-iceberg/src/test/java/com/alibaba/fluss/lake/iceberg/tiering/IcebergTieringITCase.java:64]: nit: can we separate into smaller test methods
```
    void testTiering() throws Exception { 
        testPrimaryKeyTableTiering();
        testLogTableTiering();
        testPartitionedTableTiering();
    }
```

File: fluss-lake/fluss-lake-iceberg/src/test/java/com/alibaba/fluss/lake/iceberg/tiering/IcebergTieringITCase.java [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+        Map<Long, String> partitionNameByIds = waitUntilPartitions(partitionedTablePath);
+
+        // now, write rows into partitioned table
+        TableDescriptor partitionedTableDescriptor = tableIdAndDescriptor.f1;
+        Map<String, List<InternalRow>> writtenRowsByPartition =
+                writeRowsIntoPartitionedTable(
+                        partitionedTablePath, partitionedTableDescriptor, partitionNameByIds);
+        long tableId = tableIdAndDescriptor.f0;
+
+        // wait until synced to paimon
```
Comment (MehulBatra) [inline fluss-lake/fluss-lake-iceberg/src/test/java/com/alibaba/fluss/lake/iceberg/tiering/IcebergTieringITCase.java]: nit:
change comment to iceberg

File: fluss-lake/fluss-lake-iceberg/src/test/java/com/alibaba/fluss/lake/iceberg/tiering/IcebergTieringITCase.java [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+        for (Long partitionId : partitionNameByIds.keySet()) {
+            TableBucket tableBucket = new TableBucket(tableId, partitionId, 0);
+            assertReplicaStatus(tableBucket, 3);
+        }
+
+        // now, let's check data in paimon per partition
```
Comment (MehulBatra) [inline fluss-lake/fluss-lake-iceberg/src/test/java/com/alibaba/fluss/lake/iceberg/tiering/IcebergTieringITCase.java]: ditto

File: fluss-lake/fluss-lake-iceberg/src/test/java/com/alibaba/fluss/lake/iceberg/tiering/IcebergTieringITCase.java [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+        // check data in paimon
```
Comment (MehulBatra) [inline fluss-lake/fluss-lake-iceberg/src/test/java/com/alibaba/fluss/lake/iceberg/tiering/IcebergTieringITCase.java]: ditto

File: fluss-lake/fluss-lake-iceberg/src/test/java/com/alibaba/fluss/lake/iceberg/testutils/FlinkIcebergTieringTestBase.java [hunk 5]
```diff
@@ -0,0 +1,466 @@
...
+    }
+
+    protected long createLogTable(TablePath tablePath, int bucketNum) throws Exception {
+        return createLogTable(tablePath, bucketNum, false);
+    }
+
+    protected long createLogTable(TablePath tablePath, int bucketNum, boolean isPartitioned)
+            throws Exception {
+        Schema.Builder schemaBuilder =
+                Schema.newBuilder().column("a", DataTypes.INT()).column("b", DataTypes.STRING());
```
Comment (MehulBatra) [inline fluss-lake/fluss-lake-iceberg/src/test/java/com/alibaba/fluss/lake/iceberg/testutils/FlinkIcebergTieringTestBase.java:198]: I believe we should try adding different datatypes to see how tiering would work during type conversion, for extensive coverage?

File: fluss-lake/fluss-lake-iceberg/src/test/java/com/alibaba/fluss/lake/iceberg/testutils/FlinkIcebergTieringTestBase.java [hunk 6]
```diff
@@ -0,0 +1,466 @@
...
+        return createTable(tablePath, tableBuilder.build());
+    }
+
+    protected long createPkTable(TablePath tablePath, int bucketNum) throws Exception {
+        TableDescriptor table1Descriptor =
+                TableDescriptor.builder()
+                        .schema(
+                                Schema.newBuilder()
+                                        .column("a", DataTypes.INT())
+                                        .column("b", DataTypes.STRING())
```
Comment (MehulBatra) [inline fluss-lake/fluss-lake-iceberg/src/test/java/com/alibaba/fluss/lake/iceberg/testutils/FlinkIcebergTieringTestBase.java:223]: ditto like logtable

File: fluss-lake/fluss-lake-iceberg/src/test/java/com/alibaba/fluss/lake/iceberg/tiering/IcebergTieringITCase.java [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+        TableBucket t2Bucket = new TableBucket(t2Id, 0);
+        List<InternalRow> flussRows = new ArrayList<>();
+        // write records
+        for (int i = 0; i < 10; i++) {
+            rows = Arrays.asList(row(1, "v1"), row(2, "v2"), row(3, "v3"));
+            flussRows.addAll(rows);
+            // write records
+            writeRows(t2, rows, true);
+        }
+        // check the status of replica after synced;
```
Comment (MehulBatra) [inline fluss-lake/fluss-lake-iceberg/src/test/java/com/alibaba/fluss/lake/iceberg/tiering/IcebergTieringITCase.java:153]: should we add this post writes, like we did after primaryKeyTables test, to avoid race conditions?
or is it just the nature of pk tables to reach a given snapshot state?

`waitUntilSnapshot(t2Id, 1, 0);`

File: fluss-lake/fluss-lake-iceberg/src/test/java/com/alibaba/fluss/lake/iceberg/tiering/IcebergTieringITCase.java [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+
+    private Tuple2<Long, TableDescriptor> createPartitionedTable(TablePath partitionedTablePath)
+            throws Exception {
+        TableDescriptor partitionedTableDescriptor =
+                TableDescriptor.builder()
+                        .schema(
+                                Schema.newBuilder()
+                                        .column("id", DataTypes.INT())
+                                        .column("name", DataTypes.STRING())
+                                        .column("date", DataTypes.STRING())
```
Comment (MehulBatra) [inline fluss-lake/fluss-lake-iceberg/src/test/java/com/alibaba/fluss/lake/iceberg/tiering/IcebergTieringITCase.java]: Can we add Edge case test for creating a partiton key other than String leads to InvalidTableException

File: fluss-lake/fluss-lake-iceberg/src/test/java/com/alibaba/fluss/lake/iceberg/tiering/IcebergTieringITCase.java [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+        // check data in paimon
+        checkDataInIcebergAppendOnlyTable(t2, flussRows, 0);
+
+        // then write data to the pk tables
+        // write records
+        rows = Arrays.asList(row(1, "v111"), row(2, "v222"), row(3, "v333"));
+        // write records
+        writeRows(t1, rows, false);
+
+        // check the status of replica of t2 after synced
```
Comment (MehulBatra) [inline fluss-lake/fluss-lake-iceberg/src/test/java/com/alibaba/fluss/lake/iceberg/tiering/IcebergTieringITCase.java]: I believe it should be t1, we can also add a comment for better understand why offset gets 9
// 3 initial + (3 deletes + 3 inserts) = 9

**Comment by MehulBatra** [comment]
Thank you putting in the great work, left small comments PTAL!

File: fluss-lake/fluss-lake-iceberg/src/test/java/com/alibaba/fluss/lake/iceberg/tiering/IcebergTieringITCase.java [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+                                "["
+                                        + "{\"partition_id\":0,\"bucket_id\":0,\"partition_name\":\"date=2025\",\"log_offset\":3},"
+                                        + "{\"partition_id\":1,\"bucket_id\":0,\"partition_name\":\"date=2026\",\"log_offset\":3}"
+                                        + "]");
+                    }
+                };
+
+        checkSnapshotPropertyInIceberg(partitionedTablePath, properties);
+
+        jobClient.cancel().get();
```
Comment (MehulBatra) [inline fluss-lake/fluss-lake-iceberg/src/test/java/com/alibaba/fluss/lake/iceberg/tiering/IcebergTieringITCase.java]: nit
can be used try and finally to cancel the client gracefully

File: fluss-lake/fluss-lake-iceberg/src/test/java/com/alibaba/fluss/lake/iceberg/tiering/IcebergTieringITCase.java (hunk 8 above)
Comment (luoyuxia) [reply fluss-lake/fluss-lake-iceberg/src/test/java/com/alibaba/fluss/lake/iceberg/tiering/IcebergTieringITCase.java]: I think it's not related to the IT case, it should belong to the scope of test of create Fluss table

File: fluss-lake/fluss-lake-iceberg/src/test/java/com/alibaba/fluss/lake/iceberg/tiering/IcebergTieringITCase.java (hunk 7 above)
Comment (luoyuxia) [reply fluss-lake/fluss-lake-iceberg/src/test/java/com/alibaba/fluss/lake/iceberg/tiering/IcebergTieringITCase.java:153]: There's no snapshot in t2 which is a log table.

File: fluss-lake/fluss-lake-iceberg/src/test/java/com/alibaba/fluss/lake/iceberg/testutils/FlinkIcebergTieringTestBase.java (hunk 5 above)
Comment (luoyuxia) [reply fluss-lake/fluss-lake-iceberg/src/test/java/com/alibaba/fluss/lake/iceberg/testutils/FlinkIcebergTieringTestBase.java:198]: Good point. I create a issue #1574 to track it. Let's first merge this pr.

File: fluss-lake/fluss-lake-iceberg/src/test/java/com/alibaba/fluss/lake/iceberg/tiering/IcebergTieringITCase.java (hunk 7 above)
Comment (MehulBatra) [reply fluss-lake/fluss-lake-iceberg/src/test/java/com/alibaba/fluss/lake/iceberg/tiering/IcebergTieringITCase.java:153]: Ok

**Comment by MehulBatra** [comment]
Looks good to me, merging....
